### ISSUE TITLE
Optional argument in object with variable validation issue

### ISSUE THREAD

**Issue by flmmartins** [hashicorp/terraform#29488]
### Terraform Version

```
1.0.2
```

### Terraform Configuration Files

I have a module and it has input validation like 

```terraform

# module-variable
variable "network_rule_default_action" {
  type        = string
  description = "Default action to the network rule set"
  default     = "Deny"
  validation {
    condition     = can(regex("^(Deny|Allow)$", var.network_rule_default_action))
    error_message = "Value must be Deny or Allow."
  }
}

# Definition of an acr object - a configuration can have an ACR object or notA

variable "acr" {
  type        = object({
    name                            = optional(string)
    sku_tier                        = optional(string)
    admin_enabled                   = optional(bool)
    network_rule_default_action     = optional(string)
    whitelisted_cidrs               = optional(list(string))
    whitelisted_subnets             = optional(list(string))
    retention_policy_enabled        = optional(bool)
    retention_policy_number_of_days = optional(number)
    georeplications                 = optional(list(any))
  })
  description = "ACR configuration"
  default     = {}
}

# Define var.acr
acr = {
   name = "bla"
}

# Module call

module "acr" {
  source = "../../terraform_modules/container-registry"

  name                            = var.acr.name
  sku_tier                        = var.acr.sku_tier
  admin_enabled                   = var.acr.admin_enabled
  network_rule_default_action     = var.acr.network_rule_default_action
  whitelisted_cidrs               = var.acr.whitelisted_cidrs
  whitelisted_subnets             = var.acr.whitelisted_subnets
  retention_policy_enabled        = var.acr.retention_policy_enabled
  retention_policy_number_of_days = var.acr.retention_policy_number_of_days
  georeplications                 = var.acr.georeplications
```

### Expected Behavior

My expection is that the default returned by the optional() is null. This will make the module default kick which in this case is "Deny".  Validation passes and everything works

### Actual Behavior

When I ran this during plan I get:
Error: Invalid value for variable

### Steps to Reproduce

Add   experiments = [module_variable_optional_attrs]
tf init and plan.

### References

I think the issue has to do with:

- #24142

**Comment by alisdair** [comment]
Hi @flmmartins, thanks for reporting this. I agree with your diagnosis that this is actually another instance of #24142.

> My expection is that the default returned by the optional() is null.

That seems right to me, and also seems to be happening. `optional` seems to be working as currently designed here.

> This will make the module default kick which in this case is "Deny".

This is not accurate, as binding a module input variable to a value which evaluates to `null` will **not** result in the default value being used. This is described in the linked issue, which is behaviour that we can't change (but we need to update the docs).

That said, this seems like it might be a good use of [the `defaults` function](https://www.terraform.io/docs/language/functions/defaults.html). Here's a reduced example to explain:

```hcl
variable "acr" {
  type        = object({
    name                            = string
    admin_enabled                   = optional(bool)
    network_rule_default_action     = optional(string)
  })
  description = "ACR configuration"
}

locals {
  acr = defaults(var.acr, {
    network_rule_default_action = "Deny"
  })
}

output "acr" {
  value = local.acr
}
```

With only the name specified (as in your report), this results in the following output:

```
acr = {
  "admin_enabled" = tobool(null)
  "name" = "bla"
  "network_rule_default_action" = "Deny"
}
```

Would this suit your use case?

(Note that there appears to be a bug in `defaults` preventing this from working with your full object with `optional(list(string))` attributes; I'll report that separately and link here.)

**Comment by flmmartins** [comment]
Hey @alisdair thanks for replying.

Yes, indeed I could use the defauts but I detected the optional(list) issue you mentioned and also detected a similar issue with map of strings as well so the attribute of georeplication would cause issues.

https://github.com/hashicorp/terraform/issues/28406

I found a bit strange having to define the defaults again although I have them in my module. For me it feels like duplicating code. It would be nice if we could count only with the module default in this case.

**Comment by alisdair** [comment]
Thanks for filing the enhancement ticket for module call argument defaults! I think in combination with #29511 we've covered all the issues raised by this ticket, so I'm going to close it and leave the others open. Let me know if I'm missing something and we should reopen.

**Comment by korinne** [comment]
Hi all, we recently released the [Terraform v1.3 alpha](https://releases.hashicorp.com/terraform/1.3.0-alpha20220608), which includes the ability to mark object type attributes as optional and set default values.

I'm following up with issues that have provided feedback on the previous experiment, and wanted to invite you all to try the alpha and provide any feedback on this updated design. You can learn more/add comments[ here](https://discuss.hashicorp.com/t/request-for-feedback-optional-object-type-attributes-with-defaults-in-v1-3-alpha/40550). Thank you so much in advance, and hope you like the feature!
